flang/lib/Lower/HostAssociations.cpp

```diff
@@ -160,7 +160,12 @@ class CapturedProcedure : public CapturedSymbols<CapturedProcedure> {
     if (Fortran::semantics::IsPointer(sym))
       TODO(converter.getCurrentLocation(),
            "capture procedure pointer in internal procedure");
-    return Fortran::lower::getDummyProcedureType(sym, converter);
+    mlir::Type procType = Fortran::lower::getDummyProcedureType(sym, converter);
+    const Fortran::evaluate::ProcedureDesignator procDesignator(sym);
+    if (Fortran::lower::mustPassLengthWithDummyProcedure(procDesignator,
+                                                         converter))
+      return fir::factory::getCharacterProcedureTupleType(procType);
```

Comment:
Should this `if` logic appear in `getDummyProcedureType` to avoid getting into problems elsewhere?

Reply:
Yes, that is better. Updated to do so.